File: crates/aura-daemon/src/protocol.rs
use serde::{Deserialize, Serialize};

/// Events sent from the daemon to connected UI clients over IPC.
///
/// Each event is serialized as a single JSON line (JSONL) terminated by `\n`.
/// Clients subscribe to a broadcast channel and receive all events — they can
/// filter by variant on the Swift side.
///
/// Uses internally-tagged encoding (`#[serde(tag = "type")]`) so the JSON is
/// flat — no `"data"` wrapper. Variant names are snake_case to match Swift
/// `Decodable` expectations.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum DaemonEvent {
    /// Dot color update for status bar icon.
    DotColor { color: DotColorName, pulsing: bool },

    /// A transcript line from the user or the assistant.
    Transcript {
        role: Role,
        text: String,
        done: bool,
        #[serde(default = "default_source")]
        source: String,
    },

    /// Tool execution status.
    ToolStatus {
        name: String,
        status: ToolRunStatus,
        #[serde(skip_serializing_if = "Option::is_none")]
        output: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        summary: Option<String>,
    },

    /// Status message update.
    Status { message: String },

    /// Recent session summaries for the UI empty state.
    RecentSessions { sessions: Vec<RecentSessionInfo> },

    /// The daemon is shutting down.
    Shutdown,
}

/// Dot color names sent over IPC — maps to the Swift `DotColorName` enum.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DotColorName {
    Gray,
    Green,
    Amber,
    Red,
}

/// Tool execution lifecycle status.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ToolRunStatus {
    Running,
    Completed,
    Failed,
}

/// Conversation participant role.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Role {
    User,
    Assistant,
}

/// Commands sent from UI clients to the daemon over IPC.
///
/// Each command is a single JSON line terminated by `\n`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum UICommand {
    /// Send a text message to the Gemini session (text input mode).
    SendText { text: String },

    /// Toggle microphone on/off.
    ToggleMic,

    /// Request reconnection to Gemini.
    Reconnect,

    /// Request a graceful shutdown.
    Shutdown,
}

/// A recent session summary for display in the UI.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RecentSessionInfo {
    pub session_id: String,
    pub summary: String,
    pub created_at: String,
}

fn default_source() -> String {
    "voice".into()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn transcript_serialization() {
        let event = DaemonEvent::Transcript {
            role: Role::Assistant,
            text: "Hello".into(),
            done: false,
            source: "voice".into(),
        };
        let json = serde_json::to_string(&event).unwrap();
        assert_eq!(
            json,
            r#"{"type":"transcript","role":"assistant","text":"Hello","done":false,"source":"voice"}"#
        );

        // Roundtrip
        let parsed: DaemonEvent = serde_json::from_str(&json).unwrap();
        match parsed {
            DaemonEvent::Transcript {
                role, text, done, ..
            } => {
                assert_eq!(role, Role::Assistant);
                assert_eq!(text, "Hello");
                assert!(!done);
            }
            _ => panic!("unexpected variant"),
        }
    }

    #[test]
    fn dot_color_serialization() {
        let event = DaemonEvent::DotColor {
            color: DotColorName::Green,
            pulsing: true,
        };
        let json = serde_json::to_string(&event).unwrap();
        assert_eq!(
            json,
            r#"{"type":"dot_color","color":"green","pulsing":true}"#
        );
    }

    #[test]
    fn tool_status_serialization() {
        let event = DaemonEvent::ToolStatus {
            name: "click".into(),
            status: ToolRunStatus::Running,
            output: None,
            summary: None,
        };
        let json = serde_json::to_string(&event).unwrap();
        assert_eq!(
            json,
            r#"{"type":"tool_status","name":"click","status":"running"}"#
        );
    }

    #[test]
    fn tool_status_with_output_serialization() {
        let event = DaemonEvent::ToolStatus {
            name: "click".into(),
            status: ToolRunStatus::Completed,
            output: Some("done".into()),
            summary: None,
        };
        let json = serde_json::to_string(&event).unwrap();
        assert_eq!(
            json,
            r#"{"type":"tool_status","name":"click","status":"completed","output":"done"}"#
        );
    }

    #[test]
    fn send_text_serialization() {
        let cmd = UICommand::SendText {
            text: "hello".into(),
        };
        let json = serde_json::to_string(&cmd).unwrap();
        assert_eq!(json, r#"{"type":"send_text","text":"hello"}"#);

        // Roundtrip
        let parsed: UICommand = serde_json::from_str(&json).unwrap();
        match parsed {
            UICommand::SendText { text } => assert_eq!(text, "hello"),
            _ => panic!("unexpected variant"),
        }
    }

    #[test]
    fn reconnect_serialization() {
        let cmd = UICommand::Reconnect;
        let json = serde_json::to_string(&cmd).unwrap();
        assert_eq!(json, r#"{"type":"reconnect"}"#);
    }

    #[test]
    fn shutdown_serialization() {
        let event = DaemonEvent::Shutdown;
        let json = serde_json::to_string(&event).unwrap();
        assert_eq!(json, r#"{"type":"shutdown"}"#);
    }

    #[test]
    fn status_serialization() {
        let event = DaemonEvent::Status {
            message: "Connected".into(),
        };
        let json = serde_json::to_string(&event).unwrap();
        assert_eq!(json, r#"{"type":"status","message":"Connected"}"#);
    }

    #[test]
    fn transcript_with_source_serialization() {
        let event = DaemonEvent::Transcript {
            role: Role::User,
            text: "hello".into(),
            done: true,
            source: "text".into(),
        };
        let json = serde_json::to_string(&event).unwrap();
        assert!(json.contains(r#""source":"text""#));
    }
}

File: AuraApp/Sources/Protocol.swift
import Foundation

// MARK: - Daemon -> UI Events

/// Events sent from the Rust daemon to the SwiftUI frontend over the Unix socket.
/// Each event is a single JSON line (JSONL).
enum DaemonEvent: Decodable {
    case dotColor(DotColorUpdate)
    case transcript(TranscriptUpdate)
    case toolStatus(ToolStatusUpdate)
    case status(StatusUpdate)
    case recentSessions(RecentSessionsUpdate)
    case shutdown

    private enum CodingKeys: String, CodingKey {
        case type
    }

    private enum EventType: String, Decodable {
        case dotColor = "dot_color"
        case transcript
        case toolStatus = "tool_status"
        case status
        case recentSessions = "recent_sessions"
        case shutdown
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(EventType.self, forKey: .type)

        switch type {
        case .dotColor:
            self = .dotColor(try DotColorUpdate(from: decoder))
        case .transcript:
            self = .transcript(try TranscriptUpdate(from: decoder))
        case .toolStatus:
            self = .toolStatus(try ToolStatusUpdate(from: decoder))
        case .status:
            self = .status(try StatusUpdate(from: decoder))
        case .recentSessions:
            self = .recentSessions(try RecentSessionsUpdate(from: decoder))
        case .shutdown:
            self = .shutdown
        }
    }
}

struct DotColorUpdate: Decodable {
    let color: DotColorName
    let pulsing: Bool
}

enum DotColorName: String, Decodable {
    case gray
    case green
    case amber
    case red
}

struct TranscriptUpdate: Decodable {
    let role: TranscriptRole
    let text: String
    let done: Bool
    let source: TranscriptSource

    enum TranscriptSource: Decodable, Equatable {
        case voice
        case text
        case unknown(String)

        init(from decoder: Decoder) throws {
            let container = try decoder.singleValueContainer()
            let rawValue = try container.decode(String.self)
            switch rawValue {
            case "voice": self = .voice
            case "text": self = .text
            default: self = .unknown(rawValue)
            }
        }
    }
}

enum TranscriptRole: String, Decodable {
    case user
    case assistant
}

struct ToolStatusUpdate: Decodable {
    let name: String
    let status: ToolRunStatus
    let output: String?
    let summary: String?
}

enum ToolRunStatus: String, Decodable {
    case running
    case completed
    case failed
}

struct StatusUpdate: Decodable {
    let message: String
}

struct RecentSessionsUpdate: Decodable {
    let sessions: [RecentSession]
}

struct RecentSession: Decodable, Identifiable {
    let sessionId: String
    let summary: String
    let createdAt: String

    var id: String { sessionId }

    private enum CodingKeys: String, CodingKey {
        case sessionId = "session_id"
        case summary
        case createdAt = "created_at"
    }

    /// Format the timestamp for display (e.g., "Today, 2:30 PM" or "Yesterday").
    var displayTime: String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        guard let date = formatter.date(from: createdAt) ?? ISO8601DateFormatter().date(from: createdAt) else {
            return createdAt
        }

        let calendar = Calendar.current

        if calendar.isDateInToday(date) {
            let timeFormatter = DateFormatter()
            timeFormatter.dateFormat = "h:mm a"
            return "Today, \(timeFormatter.string(from: date))"
        } else if calendar.isDateInYesterday(date) {
            let timeFormatter = DateFormatter()
            timeFormatter.dateFormat = "h:mm a"
            return "Yesterday, \(timeFormatter.string(from: date))"
        } else {
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "MMM d, h:mm a"
            return dateFormatter.string(from: date)
        }
    }
}

// MARK: - UI -> Daemon Commands

/// Commands sent from the SwiftUI frontend to the Rust daemon over the Unix socket.
/// Each command is a single JSON line (JSONL).
enum UICommand: Encodable {
    case sendText(String)
    case toggleMic
    case reconnect
    case shutdown

    private enum CodingKeys: String, CodingKey {
        case type
        case text
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .sendText(let text):
            try container.encode("send_text", forKey: .type)
            try container.encode(text, forKey: .text)
        case .toggleMic:
            try container.encode("toggle_mic", forKey: .type)
        case .reconnect:
            try container.encode("reconnect", forKey: .type)
        case .shutdown:
            try container.encode("shutdown", forKey: .type)
        }
    }
}

// MARK: - Activity Stream Model

/// A single event in the activity stream.
struct ActivityEvent: Identifiable {
    let id: UUID
    let kind: EventKind
    var text: String
    let timestamp: Date

    init(kind: EventKind, text: String) {
        self.id = UUID()
        self.kind = kind
        self.text = text
        self.timestamp = Date()
    }
}

enum EventKind: Equatable {
    case userSpeech          // 🎤 voice transcript
    case userText            // 💬 typed message
    case assistantSpeech     // 🔊 what Aura said
    case toolCall(ToolRunStatus)  // ⚡ tool execution
    case turnSeparator       // ─ ─ visual break
}
